src/schemaVisitor.ts

```diff
@@ -0,0 +1,464 @@
+  GraphQLObjectType,
+  GraphQLScalarType,
+  GraphQLSchema,
+  GraphQLUnionType,
+  Kind,
+  ValueNode,
+} from 'graphql';
+
+import {
+  getArgumentValues,
+} from 'graphql/execution/values';
+
+export type VisitableSchemaType =
+    GraphQLSchema
+  | GraphQLObjectType
+  | GraphQLInterfaceType
+  | GraphQLNamedType
+  | GraphQLScalarType
+  | GraphQLField<any, any>
+  | GraphQLArgument
+  | GraphQLUnionType
+  | GraphQLEnumType
+  | GraphQLEnumValue;
+
+const hasOwn = Object.prototype.hasOwnProperty;
+
+// Abstract base class of any visitor implementation, defining the available
+// visitor methods along with their parameter types, and providing a static
+// helper function for determining whether a subclass implements a given
+// visitor method, as opposed to inheriting one of the stubs defined here.
+export abstract class SchemaVisitor {
+  // All SchemaVisitor instances are created while visiting a specific
+  // GraphQLSchema object, so this property holds a reference to that object,
+  // in case a vistor method needs to refer to this.schema.
+  public schema: GraphQLSchema;
+
+  // Determine if this SchemaVisitor (sub)class implements a particular
+  // visitor method.
+  public static implementsVisitorMethod(methodName: string) {
+    if (! methodName.startsWith('visit')) {
+      return false;
+    }
+
+    const method = this.prototype[methodName];
+    if (typeof method !== 'function') {
+      return false;
+    }
+
+    const stub = SchemaVisitor.prototype[methodName];
+    if (method === stub) {
+      // If this.prototype[methodName] was just inherited from SchemaVisitor,
+      // then this class does not really implement the method.
+      return false;
+    }
+
+    return true;
+  }
+
+  // Concrete subclasses of SchemaVisitor should override one or more of these
+  // visitor methods, in order to express their interest in handling certain
+  // schema types/locations.
+  /* tslint:disable:no-empty */
+  public visitSchema(schema: GraphQLSchema) {}
+  public visitScalar(scalar: GraphQLScalarType) {}
+  public visitObject(object: GraphQLObjectType) {}
+  public visitFieldDefinition(field: GraphQLField<any, any>, details: {
+    objectType: GraphQLObjectType | GraphQLInterfaceType,
+  }) {}
+  public visitArgumentDefinition(argument: GraphQLArgument, details: {
+    field: GraphQLField<any, any>,
+    objectType: GraphQLObjectType | GraphQLInterfaceType,
+  }) {}
+  public visitInterface(iface: GraphQLInterfaceType) {}
+  public visitUnion(union: GraphQLUnionType) {}
+  public visitEnum(type: GraphQLEnumType) {}
+  public visitEnumValue(value: GraphQLEnumValue, details: {
+    enumType: GraphQLEnumType,
+  }) {}
+  public visitInputObject(object: GraphQLInputObjectType) {}
+  public visitInputFieldDefinition(field: GraphQLInputField, details: {
+    objectType: GraphQLInputObjectType,
+  }) {}
+  /* tslint:enable:no-empty */
+}
+
+// Generic function for visiting GraphQLSchema objects.
+export function visitSchema(
```

Comment:
I believe this `visitSchema` function is useful for any kind of schema visitor pattern, not just visiting `@directive` syntax. Since reinventing schema visitor abstractions seems pretty common (e.g. this PR and #527, and probably elsewhere), I'm hopeful that we can consolidate our efforts into one shared abstraction. That said, I don't think consolidation should be a blocker for shipping developer-facing features like `SchemaDirectiveVisitor` or schema stitching / field renaming.

Reply:
@freiksenet Looking forward to chatting about this with you and @stubailo!